appinventor/components/src/com/google/appinventor/components/runtime/Canvas.java

```diff
@@ -1264,6 +1266,116 @@ public void DrawLine(int x1, int y1, int x2, int y2) {
     view.invalidate();
   }
 
+  /**
+   * Draws a shape on the canvas.
+   * pointList should be a list contains sub-lists with two number which represents a coordinate.
+   * The first point and last point does not need to be the same. e.g. ((x1 y1) (x2 y2) (x3 y3))
+   * When fill is true, the shape will be filled.
+   *
+   * @param pointList  A list of points, should contains sub-lists with two number which represents a coordinate.
+   *                   The first point and last point does not need to be the same. e.g. ((x1 y1) (x2 y2) (x3 y3))
+   * @param fill  true for filled shape; false for shape outline
+   */
+  @SimpleFunction(description = 
+      "Draws a shape on the canvas. " +
+      "pointList should be a list contains sub-lists with two number which represents a coordinate. " +
+      "The first point and last point does not need to be the same. e.g. ((x1 y1) (x2 y2) (x3 y3)) " +
+      "When fill is true, the shape will be filled.")
+  public void DrawShape(YailList pointList, boolean fill) {
+    Path path;
+    try {
+      path = parsePath(parsePointList(pointList));
+    } catch (IllegalArgumentException e) {
+      path = null;
+    }
+    if (path == null) {
+      $form().dispatchErrorOccurredEvent(this, "DrawShape", ErrorMessages.ERROR_CANVAS_DRAW_SHAPE_BAD_ARGUMENT);
+      return;
+    }
+    path.close();
+    Paint p = new Paint(paint);
+    p.setStyle(fill ? Paint.Style.FILL : Paint.Style.STROKE);
+    view.canvas.drawPath(path, p);
+    view.invalidate();
+  }
+
+  private Path parsePath(float[][] points) throws IllegalArgumentException {
+    if (points == null) {
+      throw new IllegalArgumentException();
+    }
+    float scalingFactor = $form().deviceDensity();
+
+    Path path = new Path();
+    path.moveTo(points[0][0] * scalingFactor, points[0][1] * scalingFactor);
+    for (int i = 1; i < points.length; i++) {
```

Comment:
If one passes an empty list to `DrawShape`, an `ArrayIndexOutOfBoundsException` will be thrown here. To fix this, you should extend the if statement on line 1318 to include a check for if `pointList.size() == 0` so that an error will get reported via the form.